### PR TITLE
Core/Movement: Use time synchronisation to better interpret client timestamps

### PR DESCRIPTION
**Pull request by chaodhib**
**Changes proposed:**
- Implement [Simple Network Time Protocol](https://tools.ietf.org/html/rfc4330) (SNTP) using the `SMSG_TIME_SYNC_REQUEST `/  `CMSG_TIME_SYNC_RESPONSE` exchange. No need to read however the RFC. The mechanism is quite simple and it is explained in the code. This packet exchange initiated by the server provides to the server the following information: the difference between its own clock and the clock of the client. With that clock delta, the server can precisely interpret timestamps sent by the client.
- On top of SNTP, I implemented the approach described [HERE](https://web.archive.org/web/20180430214420/http://www.mine-control.com/zack/timesync/timesync.html). That approach is necessary to use on top of SNTP because plain SNTP is designed to work on UDP, not TCP. The fact that TCP can resend dropped packets without any warning requires extra care.
- Using the server newly found information, which is the clock delta, we can now interpret the only timestamps sent by the client to the server: The movement timestamps. Now, when movement events occur on the client, we can precisely know their time on server clock (such as the player has started to move forward, for example). And send these movement timestamps on server clock to the observing nearby players. This change should improve the visual perseption of the other players on screen.

**Target branch(es):** 3.3.5

**Issues addressed:** should improve the visual perseption of the other players on screen. Should help with https://github.com/TrinityCore/TrinityCore/issues/19271.

**Tests performed:** Build and tested IG with artificial latency and packet drop on my machine. Larger scale testing should be performed

**Known issues and TODO list:**
Nothing that I'm aware of.

### CODE REVIEW

**Comment by Aokromes** [comment]
TrinityCore/TrinityCore/src/server/game/Handlers/MovementHandler.cpp:353:162: fatal error: 
      format specifies type 'int' but the argument has type 'int64' (aka 'long')
      [-Wformat]
  ...Clock delta: %d", movementInfo.time, plrMover->m_timeSyncClockDelta);
                  ~~                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                  %ld

File: src/server/game/Handlers/MovementHandler.cpp
Comment (sucklifex) [inline src/server/game/Handlers/MovementHandler.cpp]: Please also remove this #define MOVEMENT_PACKET_TIME_DELAY 0 at line 35

**Comment by jackpoz** [comment]
what kind of tests have been performed ?

**Comment by jackpoz** [comment]
isn't that you want to revert the whole https://github.com/chaodhib/TrinityCore/commit/2cca3982d4dd87be452ab13332dd809e02d2a24b too ? You removed every use of m_clientTimeDelay but kept the class member field

**Comment by Palabola** [comment]
@Jackpoz I'm not sure, but since 5.X this delaytime can cause really stange desyncronization. Mainly because the current movement system use much less packet number then before so 1-2 wrong timed packet can cause seeable problems.

**Comment by jackpoz** [comment]
@Palabola this PR is for 335 tho

**Comment by chaodhib** [comment]
@jackpoz  `isn't that you want to revert the whole chaodhib@2cca398 too ? You removed every use of m_clientTimeDelay but kept the class member field`
Didn't saw the commit but yeah pretty much. I removed the field now.

Regarding the testing done: I havent done much so far. i just verified that the values (round trip delay and the clock delta) computed made sense. Do you have tests in mind that I could do?

File: src/server/game/Handlers/MiscHandler.cpp
Comment (tkrokli) [inline src/server/game/Handlers/MiscHandler.cpp]: cloak = a cape or drape, a sleeveless outdoor overgarment that hangs loosely from the shoulders.
clock = an instrument to indicate, keep, and co-ordinate time.

Comment (chaodhib) [inline src/server/game/Handlers/MiscHandler.cpp]: oups, thank you.

Comment (tkrokli) [inline src/server/game/Handlers/MiscHandler.cpp]: (same typo on the line below, obviously. :)

**Comment by Nawuko** [comment]
@chaodhib The world server processes never ticks faster than 50ms. So thats the smallest possible delay you can get if the client response packet gets processed on the next tick.

**Comment by chaodhib** [comment]
@Nawuko Yep, already gotten help from IRC. Thanks though.

**Comment by jackpoz** [comment]
When I ported that old commit I did few tests with more than 1 client to verify client smooth movements. I think I used Valgrind to have high update time diff to simulate lag.

**Comment by Palabola** [comment]
I preffer to use this: https://www.softperfect.com/products/connectionemulator/  for simulating high latency.

**Comment by chaodhib** [comment]
@jackpoz I tried running with multiple clients and I see no change of behavior. This PR basically brings 2 things:
- To the server, it gives a more accurate time-stamp of the moment the player started his change of movement. As far as I can tell however, other than my changes proposed in https://github.com/TrinityCore/TrinityCore/pull/18128, `movementInfo.time` is unused by the server at the moment.
- To the client, it also give the same time-stamp but as I as I can tell, **the time-stamp contained in MSG_MOVE packets at destination to the client are unused by the client**

I'm gonna elaborate. Please challenge me and tell me if I'm wrong in something. My current understanding is as follow: The server sends (at least) 2 types of messages to the client that contains a time-stamp (I'm talking about 3.3.5 branch): SMSG_SPELL_GO and MSG_MOVE packets. And in both cases, the time-stamp is completely ignored by the client. This is specially true since as far as I can tell, the client has no way of interpreting that time-stamp because the client and the server run on a different clock and as opposed to the server, the client has no way of computing the clock delta between its clock and the server's clock.

The server regularly sends to each client a SMSG_TIME_SYNC_REQUEST packet. The client answer to it with a CMSG_TIME_SYNC_RESPONSE packet which contains the client current time-stamp at the moment he received the request. This exchange of packets has only one goal: it allows the server to compute the clock delta. With that information, each time the client sends to the server a packet containing a time-stamp (so far I know only of MSG_MOVE messages), the server can accurately deduce when the event described by the message occurred (aka when the player pressed a movement key on his computer).

**Comment by tripleslash** [comment]
@chaodhib It allows the server to fix the timestamp in outgoing movement packets in order to enable proper movement extrapolation for other players (no more lagging when they mouse turn and move etc).

**Comment by Shauren** [comment]
The biggest problem with this PR is that it assumes that request processing time is 0, which definitely is not the case for packets processed for players already in world. It can range anywhere between 0 to max(WORLD_SLEEP_CONST, last_update_diff)

**Comment by chaodhib** [comment]
@tripleslash Yeah. The premise of my post above "And in both cases, the time-stamp is completely ignored by the client" is wrong. I proved it by sending to the client different MSG_MOVE packets with different time-stamp and there was a noticeable difference. If the timestamp was the server time, the character moved as soon as the client received the packet. If the timestamp was server time + 1000ms, the character moved after a delay (that I assume was 1s).

@Shauren I assume that the request processing time is 0 for only one packet: `SMSG_TIME_SYNC_REQUEST`. As soon as the client receives that packet, it just replies with a `CMSG_TIME_SYNC_RESPONSE` that contains the current time on the client clock. The time it takes to process that request is probably negligible most of the time.

**Comment by Shauren** [comment]
@chaodhib but the packet needs to be queued from worldsocket to worldsession and it must be processed, that is the time i was talking about - the time it takes to go from WorldSocket::ReadDataHandler to WorldSession::HandleTimeSyncResp (its not a simple call chain)

**Comment by jackpoz** [comment]
> If the timestamp was the server time, the character moved as soon as the client received the packet. If the timestamp was server time + 1000ms, the character moved after a delay (that I assume was 1s).

this sounds like the same research done maaaany years ago that led to the same commit I ported to TC
- https://github.com/TrinityCore/TrinityCore/commit/2cca3982d4dd87be452ab13332dd809e02d2a24b
- https://sourceforge.net/p/ascentemubackup/code/1637/

you can check the work done in the svn history of that repo if it helps.

**Comment by chaodhib** [comment]
@Shauren True. I'm not sure what to do with that (major) issue. It looks like `CMSG_PING` packets are handled in a synchronous way. Maybe the same could be done for `CMSG_TIME_SYNC_RESPONSE` packets... I really don't have much knowledge about that part..

@jackpoz 
1. Your ported commit does not take into account fluctuating latency. Mine does. Better yet, if I implement the method described [here](http://www.mine-control.com/zack/timesync/timesync.html), it is even possible to compensate for the error provoked by the TCP layer resending a dropped packet.
2. The field `_player->m_timeSyncClient` is, until now, completely unused. In other words, the method `WorldSession::HandleTimeSyncResp` has no effect and the TIME_SYNC packets exchange amounts to nothing. 
3. I'm not yet sure if this change is blizz-like. This is still a work in progress but anyway: Each time a movement packet is sent from the client to the server, it is then broadcasted to the other clients. With this PR, the broadcasted packet's timestamp will be the time the movement packet was sent by the client **on server clock**, as opposed to the time the movement packet was processed on server clock.

**Comment by Seyden** [comment]
if this is really fixing #19271 then merge itttt please :D

**Comment by Aokromes** [comment]
Conflicting files
src/server/game/Entities/Player/Player.cpp
src/server/game/Handlers/MiscHandler.cpp
src/server/game/Handlers/MovementHandler.cpp
src/server/game/Server/WorldSession.cpp

**Comment by chaodhib** [comment]
@Seyden It does not. And it's far from being ready to be merged.

**Comment by SeTM** [comment]
what about CMSG_MOVE_TIME_SKIPPED?

**Comment by chaodhib** [comment]
CMSG_MOVE_TIME_SKIPPED is sent by the client when it freezed for a certain period of time. Thus not running its internal clock during that time. The packet indicated the amount of time "lost". It's easy to see in sniffs since we know for sure that heartbeat packets are sent exactly 500ms after the last movement packet if the player is still moving. If the client freezed inbeween, the difference in the timestamp will be different then 500ms BUT only if you do not take into account the TIME_SKIPPED packet.

The implementation of the 2 TIME_SKIPPED packet is not very difficult. I will try to get into it asap (or someone else can and i will be happy to code review and merge) but it does not help with this PR.

File: src/server/game/Handlers/MovementHandler.cpp
Comment (Shauren) [inline src/server/game/Handlers/MovementHandler.cpp]: And you probably should be using packet sender's clock delta, not the currently controlled player's (if mind controlled) since the timestamp comes from controllers client (not moved player)

**Comment by dRaiNe** [comment]
I believe it's related to this: https://github.com/TrinityCore/TrinityCore/issues/10158

ArcEMU distributed/calculated the timing properly, back in WoW version 3.3.5 (or so). Maybe we could analyze its code, provided with the timing system didn't change over the years. I don't have WoW client at the moment.

**Comment by akrom23** [comment]
@chaodhib 
What do you think about that: https://github.com/lh-server/core/blob/development/src/shared/WorldPacket.h#L65
Tested on local and works good. This timer is registered when the worldsocket received packet. so we can use it instead of ``GameTime::GetMsTime()`` to properly calculate, how long took the client to respond to SMSG packet - this resolves the main problem described by @Shauren .
But we can use this timer in every movement packet. so this timer in the server time, when the packet is received by client - we can assume that the clientstamp sent in packet is processed right here (in case of no client lag)?.

**Comment by chaodhib** [comment]
@akrom23 
Thanks for the inspiration. I resumed working on this following your comment. In the end, I took the approach that we currently use to handle `CMSG_PING` packets: handle TIME_SYNC packets directly in `WorldSocket::ReadDataHandler()`.

With that, I added quite a lot of changes. Here is the new list of things provided by this PR:
- Implement SNTP using the `SMSG_TIME_SYNC_REQUEST `/  `CMSG_TIME_SYNC_RESPONSE` exchange. This packet exchange initiated by the server provides to the server the following information: the difference between its own clock and the clock of the client. With that clock delta, the server can precisely interpret timestamps sent by the client.
- On top of SNTP, I implemented the approach described [HERE](https://web.archive.org/web/20180430214420/http://www.mine-control.com/zack/timesync/timesync.html). That approach is necessary to use on top of SNTP because plain SNTP is designed to work on UDP, not TCP. The fact that TCP can resend dropped packets without any warning requires extra care.
- Using the server newly found information, which is the clock delta, we can now interpret the only timestamps sent by the client to the server: The movement timestamps. Now, when movement events occur on the client, we can precisely know their time on server clock (such as the player has started to move forward, for example). And send these movement timestamps on server clock to the observing nearby players. This change should improve the visual perseption of the other players on screen.

File: src/server/game/Server/WorldSocket.cpp
Comment (chaodhib) [inline src/server/game/Server/WorldSocket.cpp]: Feedback on this touch part would be nice

**Comment by jackpoz** [comment]
Does that SNTP stuff allow to know the timezone of the client (and save the calendar events with the right server time) ?

**Comment by chaodhib** [comment]
@jackpoz I don't think so, no. I think the TIME_SYNC packet exchange is only useful to network player movement more smoothly. It only provide a clock delta, which allows the server to interpret in server clock timestamps that were sent in client clock. That's it. There is no notion of timezone.

**Comment by chaodhib** [comment]
Let's merge this in two weeks if there is no opposition/review comments

**Comment by jackpoz** [comment]
Check no-pch build please

File: src/server/game/Server/WorldSocket.cpp
Comment (Shauren) [reply src/server/game/Server/WorldSocket.cpp]: It's not good actually. You have a race on `WorldSession::timeSyncClockDelta` in two threads - the network thread that this socket is bound to and any map update thread where movement packets are processed

I assume you need the timestamp of when this packet was received to be as accurate as possible (without the delay from packet queue) so my suggestion would be to somehow pass the timestamp to regular packet queue, possibly by extending WorldPacket class. Marking some opcodes as "high priority" would get the timestamp with `getMSTime()` while regular ones would use lower precision `GameTime::GetGameTimeMS()`

**Review by Shauren — CHANGES_REQUESTED** [review]

**Comment by Shauren** [comment]
Also codestyle (naming WorldSession members), either use our default style or just m_ that stuff for consistency with existing code - don't make a third naming variant

File: src/server/game/Server/WorldSocket.cpp
Comment (chaodhib) [reply src/server/game/Server/WorldSocket.cpp]: > It's not good actually. You have a race on WorldSession::timeSyncClockDelta in two threads - the network thread that this socket is bound to and any map update thread where movement packets are processed

Indeed. Wouldn't using `std::atomic<int64> timeSyncClockDelta` instead of `int64 timeSyncClockDelta` fix the issue though? This is the way `CMSG_PING` and `CMSG_KEEP_ALIVE` are handled.

Though, adding a received timestamp in WorldPacket (or a subclass), as suggested by you and @akrom23, would work too and would be cleaner I suppose.

File: src/server/game/Handlers/MiscHandler.cpp
Comment (chaodhib) [reply src/server/game/Handlers/MiscHandler.cpp]: This wouldn't really have any noticeable impact on the performance (this portion of the code is not in a hot path and all 3 operations are in log complexity). However, the current implementation is more readable, in my opinion. Better to keep it like this I think.

Comment (Shauren) [reply src/server/game/Handlers/MiscHandler.cpp]: I disagree with the readability part, anyone familiar with c++ should also be familiar with iterators and be able to immediately determine what is going on here

Comment (chaodhib) [reply src/server/game/Handlers/MiscHandler.cpp]: - We both agree that peformance wise, they are both the same.
- And we can probably both agree that using the public API of map is _at least_ as maintainable as using its iterator.

So both solutions are similar. Therefore, there is no reason left to spend any further time on this.

File: src/server/game/Server/WorldSocket.cpp
Comment (chaodhib) [reply src/server/game/Server/WorldSocket.cpp]: Done. The time when `CMSG_TIME_SYNC_RESP` packets are received are now recorded (before they are put into the packet queue). And then, they are handled like any other `PROCESS_THREADSAFE` packet.

**Comment by chaodhib** [comment]
nopch build and codestyle fixed. Anything else?

File: src/server/game/Server/WorldSession.cpp
Comment (Shauren) [inline src/server/game/Server/WorldSession.cpp]: Why are you initializing by creating a temporary circular_buffer object instead of directly passing "6" to _timeSyncClockDeltaQueue constructor?

File: src/server/game/Maps/Map.cpp
Comment (Shauren) [inline src/server/game/Maps/Map.cpp]: Isn't the point of updater/filter argument on WorldSession::Update to do exactly this? Execute code based on update context

Comment (chaodhib) [reply src/server/game/Maps/Map.cpp]: i could move the code inside WorldSession::UpdateForPlayersOnMap to WorldSession::Update but I would have to use a dynamic cast on the `updater` reference:
````C++
try
{
    dynamic_cast<MapSessionFilter&>(updater);
    // Send time sync packet every 10s.
    if (_timeSyncTimer > 0)
    {
        if (diff >= _timeSyncTimer)
            SendTimeSync();
        else
            _timeSyncTimer -= diff;
    }
}
catch (std::bad_cast exp)
{
}
````

Wouldn't that have a noticable performance impact?

File: src/server/game/Server/WorldSession.cpp
Comment (chaodhib) [reply src/server/game/Server/WorldSession.cpp]: fixed

File: src/server/game/Maps/Map.cpp
Comment (Shauren) [reply src/server/game/Maps/Map.cpp]: You don't need to rely on dynamic_cast, you can place the code under `updater.ProcessLogout()` - thats the other function of the filter that determines context

Comment (chaodhib) [reply src/server/game/Maps/Map.cpp]: Done. In that case, the virtual function `ProcessLogout` should probably be renamed into something more meaningful since i'm using it out of the log-out context.

Comment (Shauren) [reply src/server/game/Maps/Map.cpp]: Yes, I already renamed it long time ago in master branch https://github.com/TrinityCore/TrinityCore/blob/master/src/server/game/Server/WorldSession.h#L848

Comment (chaodhib) [reply src/server/game/Maps/Map.cpp]: Done

**Comment by chaodhib** [comment]
Merging this in 2 days.

**Comment by jackpoz** [comment]
Could you wait until after the 15th ? We are planning on releasing April TDB for 335, after that it's the perfect time to merge long standing pull requests :)

**Comment by chaodhib** [comment]
@jackpoz Sure, of course :)

**Comment by jackpoz** [comment]
TDB has been released :)

**Comment by chaodhib** [comment]
Merged. Thank you!

**Pull request merged**
